File: src/main/java/reo7sp/boardpp/util/FontFactory.java
package reo7sp.boardpp.util;

import org.newdawn.slick.SlickException;
import org.newdawn.slick.UnicodeFont;
import org.newdawn.slick.font.effects.ColorEffect;

import java.awt.Font;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by reo7sp on 1/7/14 at 4:39 PM
 */
public class FontFactory {
	private static final Map<FontSpecs, UnicodeFont> cache = new HashMap<FontSpecs, UnicodeFont>();
	private static final Font font = new Font(null, Font.PLAIN, 14);

	public static UnicodeFont newFont(int size, boolean bold, boolean italic) throws SlickException {
		final FontSpecs fontSpecs = new FontSpecs(size, bold, italic);
		UnicodeFont result = cache.get(fontSpecs);
		if (result == null) {
			result = new UnicodeFont(font, size, bold, italic);
			initFont(result);
			if (cache.size() > 4) {
				cache.clear();
			}
			cache.put(fontSpecs, result);
		}
		return result;
	}

	@SuppressWarnings("unchecked")
	private static void initFont(UnicodeFont font) throws SlickException {
		font.addAsciiGlyphs();
		font.addGlyphs(0x0400, 0x04FF);
		font.addGlyphs(0x1DC0, 0x2BFF);
		font.getEffects().add(new ColorEffect());
		font.loadGlyphs();
	}

	private static class FontSpecs {
		public final int size;
		public final boolean bold, italic;

		private FontSpecs(int size, boolean bold, boolean italic) {
			this.size = size;
			this.bold = bold;
			this.italic = italic;
		}

		@Override
		public boolean equals(Object o) {
			if (this == o) {
				return true;
			}
			if (!(o instanceof FontSpecs)) {
				return false;
			}
			FontSpecs fontSpecs = (FontSpecs) o;
			return bold == fontSpecs.bold && italic == fontSpecs.italic && size == fontSpecs.size;

		}

		@Override
		public int hashCode() {
			int result = size;
			result = 31 * result + (bold ? 1 : 0);
			result = 31 * result + (italic ? 1 : 0);
			return result;
		}
	}
}

File: src/main/java/reo7sp/boardpp/util/FontFactory.scala
/*
 Copyright 2014 Reo_SP

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
*/

package reo7sp.boardpp.util

import com.badlogic.gdx.graphics.g2d.freetype.FreeTypeFontGenerator
import com.badlogic.gdx.Gdx
import scala.collection.mutable
import com.badlogic.gdx.graphics.g2d.BitmapFont

/**
 * Created by reo7sp on 1/14/14 at 8:23 PM
 */
object FontFactory {
  val cache = new mutable.HashMap[(Int, Boolean, Boolean), BitmapFont]

  val regularFont = new FreeTypeFontGenerator(Gdx.files.classpath("font/DejaVuSans.ttf"))
  val boldFont = new FreeTypeFontGenerator(Gdx.files.classpath("font/DejaVuSans-Bold.ttf"))
  val boldObliqueFont = new FreeTypeFontGenerator(Gdx.files.classpath("font/DejaVuSans-BoldOblique.ttf"))
  val obliqueFont = new FreeTypeFontGenerator(Gdx.files.classpath("font/DejaVuSans-Oblique.ttf"))

  def newFont(size: Int, bold: Boolean, italic: Boolean): BitmapFont = {
    val fontOption = cache.get((size, bold, italic))
    if (fontOption == None) {
      val font = if (bold) {
        if (italic) {
          boldObliqueFont.generateFont(size, CharMap.all, false)
        } else {
          boldFont.generateFont(size, CharMap.all, false)
        }
      } else if (italic) {
        obliqueFont.generateFont(size, CharMap.all, false)
      } else {
        regularFont.generateFont(size, CharMap.all, false)
      }
      cache += (size, bold, italic) -> font
      font
    } else {
      fontOption.get
    }
  }

  def dispose(): Unit = {
    regularFont.dispose()
    boldFont.dispose()
    boldObliqueFont.dispose()
    obliqueFont.dispose()

    for ((specs, font) <- cache) {
      font.dispose()
    }
    cache.clear()
  }
}
